druid/src/widget/scroll.rs

```diff
@@ -203,19 +227,37 @@ impl<T: Data, W: Widget<T>> Widget<T> for Scroll<T, W> {
     fn layout(&mut self, ctx: &mut LayoutCtx, bc: &BoxConstraints, data: &T, env: &Env) -> Size {
         bc.debug_check("Scroll");
 
-        let old_size = self.clip.viewport().view_size;
+        let old_content_size = self.clip.viewport().content_size;
+        let old_self_size = self.clip.viewport().view_size;
+
         let child_size = self.clip.layout(ctx, &bc, data, env);
         log_size_warnings(child_size);
 
         let self_size = bc.constrain(child_size);
         // The new size might have made the current scroll offset invalid. This makes it valid
         // again.
         let _ = self.scroll_by(Vec2::ZERO);
-        if old_size != self_size {
+        if old_self_size != self_size {
             self.scroll_component
                 .reset_scrollbar_fade(|d| ctx.request_timer(d), env);
         }
 
+        if ((self.scroll_snap_vertical)(data, env) || (self.scroll_snap_horizontal)(data, env))
+            && self.clip.viewport().content_size != old_content_size
+        {
+            let viewport = self.clip.viewport();
+            let mut distance = Vec2::ZERO;
+            if (self.scroll_snap_vertical)(data, env) {
+                distance.y = viewport.content_size.height
+                    - (viewport.view_origin.y + viewport.view_size.height);
+            }
+            if (self.scroll_snap_horizontal)(data, env) {
+                distance.x = viewport.content_size.width
+                    - (viewport.view_origin.x + viewport.view_size.width);
+            }
```

Comment:
Alternatively, we could scroll by the viewport content size and allow the clip box to handle panning us back into the visible area. For example:
```
if (self.scroll_snap_vertical)(data, env) {
    distance.y = viewport.content_size.height;
}
if (self.scroll_snap_horizontal)(data, env) {
    distance.x = viewport.content_size.width;
}
```

Is it sloppy to give an excessive scroll distance and to rely on the clip box to pan us back into the visible area?
Or is it better to give an excessive scroll distance, because the clip box will check our view anyway so then we only do the computation once (and the computation location is centralised)?